geo/src/algorithm/bool_ops/assembly.rs

```diff
@@ -0,0 +1,335 @@
+                        debug_assert_ne!(c_idx, d_idx);
+                        snakes[c_idx].finish(d_idx);
+                        snakes[d_idx].finish(c_idx);
+                    }
+                }
+            }
+        }
+
+        let (rings, snakes_idx_map) = rings_from_snakes(&mut snakes[..]);
+
+        let mut polygons = vec![];
+        let mut children = HashMap::new();
+        for (ring_idx, ring) in rings.iter().enumerate() {
+            if ring.is_hole {
+                let mut parent_ring_idx;
+                let mut parent_snake_idx = ring.parent_snake_idx;
+                loop {
+                    parent_ring_idx = snakes_idx_map[&parent_snake_idx];
+                    let parent = &rings[parent_ring_idx];
+                    if !parent.is_hole {
+                        break;
+                    }
+                    parent_snake_idx = parent.parent_snake_idx;
+                }
+                let this_children = children.entry(parent_ring_idx).or_insert(vec![]);
+                this_children.push(ring_idx);
+            } else {
+                continue;
+            }
+        }
+
+        for (ring_idx, ring) in rings.iter().enumerate() {
+            if ring.is_hole {
+                continue;
+            }
+            let mut holes = vec![];
+            for child_idx in children.remove(&ring_idx).unwrap_or_default() {
+                let ls = split_ring(&rings[child_idx].ls, |ls| holes.push(ls));
+                holes.push(ls);
+            }
+            debug!("ext: {ext:?}", ext = ring.ls);
+            let exterior = split_ring(&ring.ls, |ls| holes.push(ls));
+            polygons.push(Polygon::new(exterior, holes));
+        }
+
+        polygons.into()
+    }
+}
+
+#[derive(Debug, Clone)]
+struct Ring<T: GeoFloat> {
+    ls: LineString<T>,
+    is_hole: bool,
+    parent_snake_idx: usize,
+}
+
+fn split_ring<T: GeoFloat, F: FnMut(LineString<T>)>(
+    ls: &LineString<T>,
+    mut cb: F,
+) -> LineString<T> {
+    let mut pts_map = BTreeMap::new();
+    let mut exterior = vec![];
+    for coord in ls.0.iter().copied() {
+        if let Some(idx) = pts_map.get(&SweepPoint::from(coord)) {
+            let new_ls: LineString<_> = exterior
+                .drain(idx..)
+                .inspect(|pt| {
+                    pts_map.remove(&SweepPoint::from(*pt)).unwrap();
+                })
+                .collect();
+            cb(new_ls);
+        }
+        pts_map.insert(SweepPoint::from(coord), exterior.len());
+        exterior.push(coord);
+    }
+    pts_map.clear();
+    LineString::from(exterior)
+}
+
+fn rings_from_snakes<T: GeoFloat>(
+    snakes: &mut [Snake<T>],
+) -> (Vec<Ring<T>>, HashMap<usize, usize>) {
+    let mut snake_idx_map = HashMap::new();
+    let mut rings = vec![];
+    for idx in 0..snakes.len() {
+        if let Some(ls) = Snake::into_ring(snakes, idx, |midx| {
+            snake_idx_map.insert(midx, rings.len());
```

Comment:
I'm having some trouble understanding `rings_from_snakes`/`Snake::into_ring`. 

Starting here, what is `midx`? I'm guessing it stands for "map index" because thats how it's being used here, but where does it come from? Diving into the code deeper (in `Snake::into_ring`)... it looks like it's the snake_idx of the "current" snake we're trying to fit into a ring. So a ring will be made up of potentially many snakes, right?

And `snake_idx_map` is the map of snake_idx to the ring_idx of the ring that the snake was used to build. Is that all correct?

Reply:
The construction of rings from segments (via snakes) is essentially the classical eulerian [circuit algorithm](https://en.wikipedia.org/wiki/Eulerian_path#Hierholzer's_algorithm).  The `midx` is an extra requirement here: we maintain some information at the "snake" level (the region above the snake) which is needed to determine which rings are exterior.  The `snakes_idx_map` tells us which snakes were made into a ring, and `midx` is the idx of the snake in the array.

I intend to update [this book](https://rmanoka.github.io/geo-crossings/book/) with an exposition of the ops implementation too, but unsure when.  We could move this into the gh-pages here for reference on the sweep and ops algorithms.